dynamiqs/sparse.py

```diff
@@ -0,0 +1,202 @@
+import functools
+
+import jax
+import jax.numpy as jnp
+from jaxtyping import Array
+
+
+class SparseDIA:
+    def __init__(self, diags: Array, offsets: tuple[int]):
+        self.offsets = offsets
+        self.diags = diags
+
+
+    def to_dense(self) -> Array:
+        """Turn any set of diagonals & offsets in a full NxN matrix."""
+        N = self.diags.shape[1]
+        out = jnp.zeros((N, N))
+        for offset, diag in zip(self.offsets, self.diags):
+            start = max(0, offset)
+            end = min(N, N + offset)
+            out += jnp.diag(diag[start:end], k=offset)
+        return out
+
+    @functools.partial(jax.jit, static_argnums=(0,1))
+    def _matmul_dense(
+        self,
+        direction: str,
+        matrix: Array,  # If matrix is a vector need to input as column vector.
+    ) -> Array:
+        """Sparse @ Dense only using information about diagonals & offsets."""
+        N = matrix.shape[0]
+        out = jnp.zeros_like(matrix)
+        for offset, diag in zip(self.offsets, self.diags):
+            start = max(0, offset)
+            end = min(N, N + offset)
+            top = max(0, -offset)
+            bottom = top + end - start
+            if direction=='left':
+                out = out.at[top:bottom, :].add(
+                    diag[start:end, None] * matrix[start:end, :]
+                )
+            else:
+                out = out.at[:, start:end].add(
+                    jnp.transpose(diag[start:end, None] * jnp.transpose(matrix[:, top:bottom]))
+                )
+        return out
+
+    @functools.partial(jax.jit, static_argnums=(0,1))
```

Comment:
I believe the proper way to handle the JIT is to have the class `SparseDIA` inherit from `eqx.Module` (see [Equinox](https://docs.kidger.site/equinox/)) and to make `self.offsets` a static field with `offsets = eqx.field(static=True)`.

This way, both `self` and `matrix` will be dynamic argnums (not static). So you can fully remove the decorator here.